### PR TITLE
feat: Handle Class Imbalance using SMOTE in Portuguese Bank Marketing Pipeline (#2049)

### LINKED ISSUE #2049: Handle Class Imbalance Using SMOTE
# Handle Class Imbalance Using SMOTE 
### Problem

The current implementation trains the classification model on the original dataset without addressing class imbalance. In the Portuguese Bank Marketing dataset, the target variable (`y`) is typically skewed toward the "no" class, which can cause the model to favor the majority class and reduce its ability to correctly identify the minority ("yes") class.

As a result, overall accuracy may appear high while recall and F1-score for the minority class remain poor.

### Proposed Solution

Introduce a class balancing pipeline using **SMOTE (Synthetic Minority Over-sampling Technique)** after splitting the data into training and testing sets.

Suggested workflow:

* Perform data preprocessing.
* Split the dataset into training and testing sets.
* Apply **SMOTE only on the training set** to avoid data leakage.
* Retrain the existing classification models using the balanced training data.
* Compare performance before and after resampling.

### Evaluation Metrics

Evaluate both the original and SMOTE-based models using:

* Accuracy
* Precision
* Recall
* F1-score
* ROC-AUC Score
* Confusion Matrix

This comparison will help determine whether resampling improves the model's ability to identify customers likely to subscribe to a term deposit.

### Expected Outcome

* Better detection of the minority class.
* Improved Recall and F1-score.
* More reliable evaluation of model performance on imbalanced data.
* A stronger and more practical classification pipeline.

### Possible Enhancements

* Compare SMOTE with other resampling techniques such as:

  * Random Over Sampling
  * Random Under Sampling

* Include visualizations of class distribution before and after resampling.

### Benefits

* Improves model robustness.
* Demonstrates best practices for handling imbalanced classification problems.
* Makes the project more suitable for real-world banking datasets where positive responses are relatively rare.

### PR DESCRIPTION
**Pull request by Rakshak05**
Closes #2049.

#### **Description**

This PR addresses the class imbalance in the Portuguese Bank Marketing classification pipeline by introducing class balancing techniques (SMOTE, Random Over-sampling, and Random Under-sampling) on the training set. It evaluates and compares their performance against the original unbalanced model to ensure better detection of the minority class (y = "yes").

#### **Changes Made**
- **Updated Dependencies**: Added `imbalanced-learn` to `requirements.txt` for resampling support.
- **Resampling Implementation**: Integrated `SMOTE`, `RandomOverSampler` (ROS), and `RandomUnderSampler` (RUS) only on the preprocessed training set to prevent data leakage.
- **Model Comparison**: Trained all 7 classification models across the 4 scenarios (Original, SMOTE, ROS, RUS) and compared them using Accuracy, Precision, Recall, F1-Score, and ROC-AUC.
- **Aesthetics & Visualization**:
  - Plotted bar charts comparing the class distributions before and after resampling.
  - Plotted metric comparison charts showing Recall and F1-score improvements.
- **Pipeline Deployment**: Trained the final Gradient Boosting model on the SMOTE-balanced training set and successfully saved the pipeline to `model/gradient_boosting_model.pkl`.

#### **Evaluation Metrics (Gradient Boosting)**

| Scenario | Accuracy | Precision | Recall | F1-Score | ROC-AUC |
| :--- | :---: | :---: | :---: | :---: | :---: |
| **Original (Unbalanced)** | **90.83%** | 65.37% | 46.03% | 54.02% | **92.94%** |
| **SMOTE (Balanced)** | 89.63% | 54.34% | 70.98% | **61.56%** | 92.51% |
| **ROS (Balanced)** | 85.33% | 43.60% | 86.67% | 58.02% | 92.91% |
| **RUS (Balanced)** | 84.33% | 41.88% | **87.52%** | 56.65% | 92.64% |

*The SMOTE-balanced model yielded the best overall balance between Precision and Recall with the highest F1-Score of **61.56%** (up from 54.02%) and a significantly higher Recall of **70.98%** (up from 46.03%).*

#### **Verification**
The saved pipeline was successfully tested and verified using `predict.py`:
```bash
python predict.py --data '{"age": 30, "job": "management", "marital": "married", "education": "tertiary", "default": "no", "balance": 1500, "housing": "yes", "loan": "no", "contact": "cellular", "day": 5, "month": "may", "duration": 250, "campaign": 1, "pdays": -1, "previous": 0, "poutcome": "unknown"}'

### CODE REVIEW

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]
remove all pngs and make a pdf explaining all of them @Rakshak05 
and update the readme of what changes u made

**Comment by Sibam-Paul** [comment]
@Rakshak05 feel free to make the change I will merge it

**Review by Sibam-Paul — APPROVED** [review]
LGTM! @Rakshak05
